logprep/framework/pipeline.py

```diff
@@ -120,100 +200,106 @@ def __init__(
+    @cached_property
+    def _metrics_exposer(self):
+        return MetricExposer(
             self._logprep_config.get("metrics", {}),
             self._metric_targets,
             self._shared_dict,
             self._lock,
         )
-        self.metrics = self.PipelineMetrics(
+
+    @cached_property
+    def metrics(self):
+        """The pipeline metrics object"""
+        return self.PipelineMetrics(
             input=self._input.metrics, output=self._output.metrics, labels=self._metric_labels
         )
 
-    def _build_pipeline(self):
-        self._logger.debug(f"Building '{current_process().name}'")
-        self._pipeline = tuple(
+    @cached_property
+    def _pipeline(self):
+        self._logger.debug(f"Building '{self._process_name}'")
+        pipeline = tuple(
             (self._create_processor(entry) for entry in self._logprep_config.get("pipeline"))
         )
-        self._logger.debug(f"Finished building pipeline ({current_process().name})")
-
-    def _create_processor(self, entry: dict) -> "Processor":
-        processor_name = list(entry.keys())[0]
-        entry[processor_name]["metric_labels"] = self._metric_labels
-        processor = Factory.create(entry, self._logger)
-        processor.setup()
-        self.metrics.pipeline.append(processor.metrics)
-        self._logger.debug(f"Created '{processor}' processor ({current_process().name})")
-        return processor
-
-    def _create_connectors(self):
-        self._logger.debug(f"Creating connectors ({current_process().name})")
-        self._create_input_connector()
-        self._create_output_connector()
-        self._output.input_connector = self._input
-        self._logger.debug(
-            f"Created connectors -> input: '{self._input.describe()}',"
-            f" output -> '{self._output.describe()}' ({current_process().name})"
-        )
-        self._input.setup()
-        self._output.setup()
-        if hasattr(self._input, "server"):
-            while self._input.server.config.port in self._used_server_ports:
-                self._input.server.config.port += 1
-            self._used_server_ports.update({self._input.server.config.port: current_process().name})
-        self._logger.debug(f"Finished creating connectors ({current_process().name})")
+        self._logger.debug(f"Finished building pipeline ({self._process_name})")
+        return pipeline
 
-    def _create_output_connector(self):
+    @cached_property
+    def _output(self):
         output_connector_config = self._logprep_config.get("output")
         connector_name = list(output_connector_config.keys())[0]
         output_connector_config[connector_name]["metric_labels"] = self._metric_labels
-        self._output = Factory.create(output_connector_config, self._logger)
+        return Factory.create(output_connector_config, self._logger)
 
-    def _create_input_connector(self):
+    @cached_property
+    def _input(self):
```

Comment:
Can you type the cached properties please? My IDE is giving me now 23 more warnings than the current main-state. I think many of these warnings can be prevented if the typing would be more strict. Most of the times those warnings are something like `Unresolved attribute reference 'store' for class 'Component' ` for the line `self._output.store(event)`. Yes the output does have a `store` method, and output is a subclass of `Component` but if we strictly specify that the attribute `_output` is of type `Output` then we can prevent these warnings.

Reply:
DONE